Create fastpath backend context manager, similar to SDPA kernel backend manager

mikekgfb opened the pull request:
Summary: 

1. create context manager infrastructure with aten-managed context with room for 64 settings
   - as Python builtin for Python eager and torch.compile
   - as TS builtins guarded by torch.hit.is_scripting() for legacy compatibility with TS legacy code. (since "normal" Python builtins that are part of the Pytorch distro are not accessible in TS.)
2. Ability to start with True or False initialization (we always init the state to 0, but will invert some bits based on 
       desired startup/default polarity of a flag)
3. Create fastpath backend context manager using 4 context bits from 1 
       (similar to SDPA kernel backend manager, to give users instant familiarity with the mechanism)
4. Use context manager in FSDP test code (this is prep for a future update that broadens use of fastpath, and would
      otherwise break this FSDP test when the executions are performed using different kernels, showing divergence in
      the test, not thru an error, but the use of different kernels, with different FP rounding characteristics etc)

Test Plan: sandcastle, github

Differential Revision: D48325593



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

drisspg commented on aten/src/ATen/Context.cpp:
```diff
@@ -117,6 +117,17 @@ void Context::setSDPUseMath(bool e) {
   enabled_mathSDP = e;
 }
 
+void Context::assignGlobalCtx(int pos, bool e){
```
nit: I think some more description, and maybe longer argument names could be helpful for context

mikekgfb replied:
Done

drisspg commented on torch/csrc/Module.cpp:
```diff
@@ -604,6 +604,61 @@ PyObject* THPModule_userEnabledMathSDP(PyObject* _unused, PyObject* noargs) {
   else
     Py_RETURN_FALSE;
 }
+PyObject* THPModule_setGlobalCtx(PyObject* _unused, PyObject* arg) {
```
I think some tests exercising the error cases could also be good

mikekgfb replied:
Added test_opctl and some fastpath runs of actual models.

drisspg commented:
Overall I think looks good just a few comments about testing

davidberard98 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -181,6 +181,45 @@ static const std::vector<OperatorGeneratorArgs> opGenArgs{
+          pop(stack, ctx_no);
+          at::globalContext().assignGlobalCtx(ctx_no, true);
+        },
+        aliasAnalysisFromSchema()),
```
I think these should actually be `aliasAnalysisConservative()` because they have side effects. Otherwise, TorchScript might try to optimize by moving these ops around the graph.

I think it should be safe to leave `_GetGlobalCtx` as being `aliasAnalysisFromSchema()`

mikekgfb replied:
Done

mikaylagawarecki reviewed (commented):
Hey @mikekgfb, I haven't done an in-depth review of this but I have two broad questions

 I take it that the intent is for this context manager to be able to disable the fastpaths for `nn.MHA` and `nn.Transformer`. However, while the flags in the context manager for SDPA toggle between the 3 backends that implement SDPA (math, mem-eff and flash), it appears to me like the flags in this context manager might be a bit more nuanced.

1) is the term `ATFP` widely used to refer to the combination of `nn.MHA`/`nn.Transformer`? I am curious about the rationale for combining these into one single context manager as well as whether the naming will make this discoverable?


2) Wanted to understand the meanings of the kwargs to this context manager. Granted that this is probably intended to be a tool for power users who have good context on this but I think it is important that we clearly establish the rationale/use-cases for each the arguments to understand the design here. 


- `math`: mirrors the context manager for sdpa 
- `enable_nested_tensor`: gives the ability to override the `TransformerEncoder(enable_nested_tensor)` flag given at construction time during runtime (
- `enable_mha` gives the ability to disable the MHA sparsity fast path during runtime (doesn't have a corresponding flag in MHA constructor)
- `enable_encoder` disables the sparsity fast path for `TransformerEncoderLayer` at runtime (seems like naming might be confusing here)

What is the rationale for including both flags that override arguments to constructors of certain modules as well as other flags that directly disable the fast path for certain modules?

mikekgfb commented:
> What is the rationale for including both flags that override arguments to constructors of certain modules as well as other flags that directly disable the fast path for certain modules?

All flags disable the mode dynamically, starting and ending at the scope of the context manager, regardless of whatever settings.  IN some ways you might think of that similar to how SDPA context manager works -- it enables or disables specific kernels, but they can also be not accessible on some hardware etc. So context manager is one of many decision criteria, but it's a surefire way to disable if

1 - it's broken

2 - performance is not good

3 - numerical equivalence is important (this was the original starting point for building this -- https://github.com/pytorch/pytorch/pull/106668 enables fastpath for inputs that are `not batch_first`.  However, by doing this, it breaks the test_fsdp_core which runs the model with model.eval and model w/ no_grad() and exects them to be bit-accurate the same.  This only works if we use the same kernels)

4 - users don't want the fastpath for other reasons, e.g., the recent https://github.com/pytorch/pytorch/pull/106824.  We can keep adding additional conditions for each case, or just have a framework for user control.  

The naming is  an interesting question - fp stands for the (inference) "fastpath".  We have captured all these features as Accelerated Transformer (previously known as Better Transformer) - so that's where the at comes from to make allowance that "fastpath" isn't super descriptive.

I think they fit reasonably well together, and give users control over what e2e consists of interlocking features.  I can see other ways to dividing them, or I could also see the case for creating it as _atfp_* to introduce it first as a non-public interface to gain some experience. Wdyt?

mikekgfb commented:
@Mikayla ***@***.***> I’ll be appreciate to get your suggestion for naming.  I agree that ATFP isn’t particularly obvious - I was concerned that fastpath alone was too generic (there are many operators that might have a fastpath. Totally not clear it refers to transformers. We  pulso call it the longer name transformer_fastpath_

You asked about why some operators were disabling construction time settings. I think when we first introduced these settings, we wanted to give users control. Since then it turns out that this may not be enough, i.e., a model might be constructed once, but then users may want to run it without the fastpath.

Two examples:
A - A recent change disabled fastpath diff scripted models because on edge devices, they don’t include the operator. Takeaway: they build the model with fastpath (or received one that was so built), and the current “fix” disables the fastest for every torchscript user of i understand things correctly!

B - For the FSDP unit tests, the model is built with enable nested tensor, but accuracy checks are performed to check bitwise accuracy. That is bound to fail because it will trigger different rounding even if the execution itself is mathematically equivalent.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: mikaylagawarecki ***@***.***>
Sent: Friday, August 18, 2023 8:16:27 AM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Mention ***@***.***>
Subject: Re: [pytorch/pytorch] Create fastpath backend context manager, similar to SDPA kernel backend manager (PR #107163)

@mikaylagawarecki commented on this pull request. Hey @mikekgfb, I haven't done an in-depth review of this but I have two broad questions I take it that the intent is for this context manager to be able to disable the fastpaths for MHA and
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

@mikaylagawarecki commented on this pull request.

Hey @mikekgfb<https://github.com/mikekgfb>, I haven't done an in-depth review of this but I have two broad questions

I take it that the intent is for this context manager to be able to disable the fastpaths for MHA and Transformer. However, while the flags in the context manager for SDPA clearly toggle between the 3 backends, it appears to me like the flags in this context manager might be a bit more nuanced.

  1.  is the term atfp widely used to refer to the combination of nn.MHA/`nn.Transformer? I am curious about the rationale for combining these into one single context manager as well as whether the naming will make this discoverable?

  2.  Wanted to understand the meanings of the kwargs to this context manager. Granted that this is probably intended to be a tool for power users who have good context on this but I think it is important that we clearly establish the meanings of the arguments to understand the design here.

  *   math is clear to me and mirrors the context manager for sdpa
  *   enable_nested_tensor: gives the ability to override the TransformerEncoder(enable_nested_tensor) flag given at construction time during runtime?
  *   enable_mha gives the ability to disable the MHA sparsity fast path during runtime (doesn't have a corresponding flag in MHA constructor)
  *   enable_encoder disables the sparsity fast path for TransformerEncoderLayer at runtime (seems like naming might be confusing here)

What is the rationale for including both flags that override arguments to the constructor as well as other flags that directly disable the fast path?

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/107163#pullrequestreview-1583302961>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHLN62OE554IWXPFXALXVZUYXANCNFSM6AAAAAA3QDUY5Q>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Commit:
Create fastpath backend context manager, similar to SDPA kernel backend manager (#107163)

Summary:

Create fastpath backend context manager, similar to SDPA kernel backend manager
ghstack-source-id: 208858046
exported-using-ghexport

Test Plan: sandcastle, github

Reviewed By: osalpekar

Differential Revision: D48325593
